{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.insert(0, '../scripts') # ! This wont work now as i moved parse_motifcen  to source\n",
    "from parse_motifcentral_json import MOTIFCENTRAL \n",
    "from get_search_results_from_ENCODE import ENCODE_search "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['modelSettings', 'coefficients', 'metadata'])"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# MOTIFCENTRAL is a list of JSON objects. Each JSON object corresponds to a single motif.\n",
    "MOTIFCENTRAL[0].keys()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MOTIFCENTRAL[0]['modelSettings'].keys() = dict_keys(['letterOrder', 'bindingModes', 'letterComplement'])\n",
      "MOTIFCENTRAL[0]['coefficients'].keys() = dict_keys(['bindingModes'])\n",
      "MOTIFCENTRAL[0]['metadata'].keys() = dict_keys(['fit_id', 'factors', 'experiments'])\n"
     ]
    }
   ],
   "source": [
    "print(f\"{MOTIFCENTRAL[0]['modelSettings'].keys() = }\") \n",
    "print(f\"{MOTIFCENTRAL[0]['coefficients'].keys() = }\") \n",
    "print(f\"{MOTIFCENTRAL[0]['metadata'].keys() = }\") "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. What are all the metadata.factors.gene_symbol in MOTIFCENTRAL?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MOTIFCENTRAL[0]['metadata']['factors'][0].keys() = dict_keys(['gene_symbol', 'gene_name', 'tax_id', 'annotation'])\n"
     ]
    }
   ],
   "source": [
    "print(f\"{MOTIFCENTRAL[0]['metadata']['factors'][0].keys() = }\") # tax_id gives information about the organism. 10090 = Musmusculus "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'gene_symbol': 'Tfe3',\n",
       "  'gene_name': 'transcription factor E3',\n",
       "  'tax_id': 10090,\n",
       "  'annotation': [['ENSEMBL_GENE', 'ENSMUSG00000000134'],\n",
       "   ['ENTREZ', '209446'],\n",
       "   ['MGI', '98511'],\n",
       "   ['PFAM', 'PF00010'],\n",
       "   ['PFAM', 'PF11851'],\n",
       "   ['PFAM', 'PF15951'],\n",
       "   ['SWISS_PROT', 'Q64092']]}]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MOTIFCENTRAL[0]['metadata']['factors']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_set = {(lambda fit: fit['metadata']['factors'][0]['gene_symbol'])(fit) for fit in MOTIFCENTRAL}\n",
    "gene_list = list([(lambda fit: fit['metadata']['factors'][0]['gene_symbol'])(fit) for fit in MOTIFCENTRAL])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Lenght of MOTIFCENTRAL = 874 \n",
      " Lenght of gene_set = 824 \n",
      " Lenght of gene_list = 874 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(f\"Lenght of MOTIFCENTRAL = {len(MOTIFCENTRAL)} \\n Lenght of gene_set = {len(gene_set)} \\n Lenght of gene_list = {len(gene_list)} \\n\") "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So we have repeats. Perhaps the same TF in different organisms? TF + Organism should be a unique identifier. The ENTREZ ID Accounts for this.\n",
    "So how do I find ChIP datasets for a given ENTREZ ID? I need to use the metadata to find my relevant genomic assay."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "2. What is the range of number of binding modes? MOTIFCENTRAL[0]['modelSettings']['bindingModes'] or MOTIFCENTRAL[0]['coefficients']['bindingModes']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'fitLogActivity': True,\n",
       "  'flankLength': 0,\n",
       "  'singleStrand': False,\n",
       "  'size': 12,\n",
       "  'dinucleotideDistance': 0,\n",
       "  'componentName': 'Binding mode 1',\n",
       "  'positionBias': False,\n",
       "  'modifications': []}]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MOTIFCENTRAL[0]['modelSettings']['bindingModes']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{1}"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "number_of_BM_range = {(lambda fit: len(fit['coefficients']['bindingModes']))(fit) for fit in MOTIFCENTRAL}\n",
    "number_of_BM_range\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Okay so we always have one BM. That makes things easier."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['mononucleotide', 'activity', 'dinucleotide', 'positionBias', 'modifications'])"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MOTIFCENTRAL[0]['coefficients']['bindingModes'][0].keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{7227, 9606, 10090, 94885}"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tax_id_set # So we only have 4 organisms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "tax_id_set = {(lambda fit: fit['metadata']['factors'][0]['tax_id'])(fit) for fit in MOTIFCENTRAL}\n",
    "tax_id_dict = {\n",
    "                7227:\"Drosophila+melanogaster\", \n",
    "                9606:\"Homo+sapiens\",\n",
    "                10090:\"Mus+musculus\",\n",
    "                94885:\"Pantherophis+guttatus\"}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# A function that returns inforamtion about fit that I can use to query the ENCODE DB\n",
    "fit_info = lambda fit: (fit['metadata']['fit_id'], fit['metadata']['factors'][0]['gene_symbol'], fit['metadata']['factors'][0]['tax_id'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TFE3 for Mus+musculus had 0.\n",
      "GATA3 for Mus+musculus had 0.\n",
      "EGR1 for Homo+sapiens had 14.\n",
      "RXRA for Homo+sapiens had 7.\n",
      "SOX17 for Homo+sapiens had 0.\n",
      "SP4 for Mus+musculus had 0.\n",
      "RARG for Mus+musculus had 0.\n",
      "GSC for Drosophila+melanogaster had 0.\n",
      "RARG for Homo+sapiens had 1.\n",
      "VSX1 for Mus+musculus had 0.\n",
      "MAFB for Mus+musculus had 0.\n",
      "MAFB for Homo+sapiens had 0.\n",
      "MAX for Homo+sapiens had 23.\n",
      "CEBPG for Homo+sapiens had 5.\n",
      "DBP for Homo+sapiens had 1.\n",
      "NFIL3 for Homo+sapiens had 1.\n",
      "DBP for Mus+musculus had 0.\n",
      "ATF4 for Mus+musculus had 0.\n",
      "SOX5 for Homo+sapiens had 1.\n",
      "MLX for Mus+musculus had 0.\n",
      "NR2F6 for Mus+musculus had 0.\n",
      "SREBF1 for Mus+musculus had 0.\n",
      "SREBF2 for Homo+sapiens had 5.\n",
      "HOMEZ for Homo+sapiens had 2.\n",
      "PKNOX1 for Homo+sapiens had 4.\n",
      "EMS for Drosophila+melanogaster had 0.\n",
      "BSH for Drosophila+melanogaster had 0.\n",
      "BCD for Drosophila+melanogaster had 0.\n",
      "DLL for Drosophila+melanogaster had 0.\n",
      "HR3 for Drosophila+melanogaster had 0.\n",
      "GCM2 for Drosophila+melanogaster had 0.\n",
      "TLL for Drosophila+melanogaster had 0.\n",
      "NFE2 for Homo+sapiens had 4.\n",
      "EVE for Drosophila+melanogaster had 0.\n",
      "ABD-B for Drosophila+melanogaster had 0.\n",
      "NR3C1 for Mus+musculus had 0.\n",
      "RFX3 for Mus+musculus had 0.\n",
      "CREB3L1 for Homo+sapiens had 1.\n",
      "HIC1 for Mus+musculus had 0.\n",
      "MNT for Homo+sapiens had 7.\n",
      "MAFK for Homo+sapiens had 9.\n",
      "NR2F6 for Homo+sapiens had 5.\n",
      "POU2F2 for Mus+musculus had 0.\n",
      "HNF4A for Mus+musculus had 0.\n",
      "CEBPA for Homo+sapiens had 1.\n",
      "ZNF777 for Homo+sapiens had 2.\n",
      "NFKB1 for Mus+musculus had 0.\n",
      "TTK for Drosophila+melanogaster had 0.\n",
      "SOX2 for Homo+sapiens had 0.\n",
      "ALX3 for Homo+sapiens had 0.\n",
      "BHLHE22 for Homo+sapiens had 0.\n",
      "SOX6 for Homo+sapiens had 3.\n",
      "FOXC1 for Homo+sapiens had 1.\n",
      "EMX2 for Homo+sapiens had 0.\n",
      "EOMES for Homo+sapiens had 0.\n",
      "E2F1 for Homo+sapiens had 8.\n",
      "HOXA2 for Mus+musculus had 0.\n",
      "LZ for Drosophila+melanogaster had 0.\n",
      "GBX2 for Homo+sapiens had 0.\n",
      "GBX2 for Mus+musculus had 0.\n",
      "HNF1A for Homo+sapiens had 2.\n",
      "IRF7 for Homo+sapiens had 0.\n",
      "NR2C2 for Homo+sapiens had 8.\n",
      "MGA for Homo+sapiens had 2.\n",
      "USP for Drosophila+melanogaster had 0.\n",
      "JDP2 for Mus+musculus had 0.\n",
      "NKX3-1 for Homo+sapiens had 1.\n",
      "MEF2D for Homo+sapiens had 2.\n",
      "ISX for Homo+sapiens had 1.\n",
      "POU3F2 for Homo+sapiens had 0.\n",
      "OTX2 for Homo+sapiens had 1.\n",
      "POU3F4 for Homo+sapiens had 0.\n",
      "RXRG for Homo+sapiens had 0.\n",
      "ZFP740 for Mus+musculus had 0.\n",
      "TFAP2B for Homo+sapiens had 1.\n",
      "TFEB for Homo+sapiens had 0.\n",
      "TCF3 for Homo+sapiens had 3.\n",
      "SOX21 for Homo+sapiens had 0.\n",
      "VAX2 for Homo+sapiens had 0.\n",
      "TFAP2A for Mus+musculus had 0.\n",
      "SAGE for Drosophila+melanogaster had 0.\n",
      "HOXA11 for Mus+musculus had 0.\n",
      "ZNF740 for Homo+sapiens had 4.\n",
      "CUX1 for Homo+sapiens had 4.\n",
      "FKH for Drosophila+melanogaster had 0.\n",
      "HOXB3 for Homo+sapiens had 0.\n",
      "PAX6 for Homo+sapiens had 0.\n",
      "SOX18 for Homo+sapiens had 1.\n",
      "BARHL1 for Mus+musculus had 0.\n",
      "LHX4 for Mus+musculus had 0.\n",
      "DLX2 for Homo+sapiens had 0.\n",
      "EY for Drosophila+melanogaster had 0.\n",
      "ZIC3 for Mus+musculus had 0.\n",
      "NFKB1 for Homo+sapiens had 0.\n",
      "TRP53 for Mus+musculus had 0.\n",
      "SP1 for Homo+sapiens had 15.\n",
      "HNG1 for Drosophila+melanogaster had 0.\n",
      "DVE for Drosophila+melanogaster had 0.\n",
      "CG31388 for Drosophila+melanogaster had 1.\n",
      "PHDP for Drosophila+melanogaster had 1.\n",
      "CNC for Drosophila+melanogaster had 0.\n",
      "MAF-S for Drosophila+melanogaster had 0.\n",
      "TOE for Drosophila+melanogaster had 0.\n",
      "EN for Drosophila+melanogaster had 0.\n",
      "SLP2 for Drosophila+melanogaster had 0.\n",
      "REPO for Drosophila+melanogaster had 0.\n",
      "CREBB for Drosophila+melanogaster had 0.\n",
      "FD59A for Drosophila+melanogaster had 0.\n",
      "SIDPN for Drosophila+melanogaster had 0.\n",
      "ATF-2 for Drosophila+melanogaster had 0.\n",
      "CG9876 for Drosophila+melanogaster had 1.\n",
      "ZEN2 for Drosophila+melanogaster had 0.\n",
      "KR for Drosophila+melanogaster had 0.\n",
      "LBE for Drosophila+melanogaster had 0.\n",
      "HNF4 for Drosophila+melanogaster had 0.\n",
      "OC for Drosophila+melanogaster had 0.\n",
      "CG17802 for Drosophila+melanogaster had 1.\n",
      "SLOU for Drosophila+melanogaster had 0.\n",
      "HR39 for Drosophila+melanogaster had 0.\n",
      "HR38 for Drosophila+melanogaster had 0.\n",
      "NK7.1 for Drosophila+melanogaster had 1.\n",
      "DA for Drosophila+melanogaster had 0.\n",
      "FOXP for Drosophila+melanogaster had 0.\n",
      "MYB for Drosophila+melanogaster had 0.\n",
      "BRK for Drosophila+melanogaster had 0.\n",
      "PNT for Drosophila+melanogaster had 0.\n",
      "HR96 for Drosophila+melanogaster had 0.\n",
      "NFI for Drosophila+melanogaster had 0.\n",
      "SCRT for Drosophila+melanogaster had 0.\n",
      "HKB for Drosophila+melanogaster had 0.\n",
      "CG4424 for Drosophila+melanogaster had 1.\n",
      "ERR for Drosophila+melanogaster had 1.\n",
      "GSB-N for Drosophila+melanogaster had 0.\n",
      "EIP75B for Drosophila+melanogaster had 0.\n",
      "EIP74EF for Drosophila+melanogaster had 0.\n",
      "OVO for Drosophila+melanogaster had 0.\n",
      "TJ for Drosophila+melanogaster had 0.\n",
      "DPN for Drosophila+melanogaster had 0.\n",
      "E(SPL)MDELTA-HLH for Drosophila+melanogaster had 0.\n",
      "NOM for Drosophila+melanogaster had 0.\n",
      "BLIMP-1 for Drosophila+melanogaster had 0.\n",
      "AWH for Drosophila+melanogaster had 0.\n",
      "ATF3 for Drosophila+melanogaster had 0.\n",
      "RETN for Drosophila+melanogaster had 0.\n",
      "HSF for Drosophila+melanogaster had 0.\n",
      "CG10654 for Drosophila+melanogaster had 1.\n",
      "EXEX for Drosophila+melanogaster had 0.\n",
      "ZEB1 for Mus+musculus had 0.\n",
      "PRD for Drosophila+melanogaster had 0.\n",
      "DLX1 for Homo+sapiens had 0.\n",
      "HEY2 for Homo+sapiens had 0.\n",
      "HOXC10 for Mus+musculus had 0.\n",
      "LHX8 for Mus+musculus had 0.\n",
      "NRF1 for Homo+sapiens had 12.\n",
      "PITX1 for Homo+sapiens had 1.\n",
      "POU2F1 for Homo+sapiens had 1.\n",
      "PAX5 for Homo+sapiens had 4.\n",
      "PRRX2 for Mus+musculus had 0.\n",
      "NKX3-2 for Homo+sapiens had 0.\n",
      "HMX3 for Homo+sapiens had 0.\n",
      "RARB for Mus+musculus had 0.\n",
      "LMX1A for Homo+sapiens had 0.\n",
      "MSX3 for Mus+musculus had 0.\n",
      "FOXL1 for Homo+sapiens had 0.\n",
      "TRL for Drosophila+melanogaster had 0.\n",
      "SOX102F for Drosophila+melanogaster had 0.\n",
      "CRC for Drosophila+melanogaster had 0.\n",
      "B-H2 for Drosophila+melanogaster had 1.\n",
      "TIN for Drosophila+melanogaster had 0.\n",
      "SLP1 for Drosophila+melanogaster had 0.\n",
      "VRI for Drosophila+melanogaster had 0.\n",
      "DLIP3 for Drosophila+melanogaster had 0.\n",
      "REL for Drosophila+melanogaster had 0.\n",
      "SENS for Drosophila+melanogaster had 0.\n",
      "PITA for Drosophila+melanogaster had 0.\n",
      "CG15696 for Drosophila+melanogaster had 1.\n",
      "NFE2L1 for Homo+sapiens had 3.\n",
      "CG7786 for Drosophila+melanogaster had 1.\n",
      "SV for Drosophila+melanogaster had 0.\n",
      "BARX1 for Homo+sapiens had 0.\n",
      "ALX4 for Homo+sapiens had 0.\n",
      "SMAD4 for Mus+musculus had 0.\n",
      "FOXG1 for Homo+sapiens had 0.\n",
      "ETV1 for Homo+sapiens had 1.\n",
      "FOXG1 for Mus+musculus had 0.\n",
      "CREB3L2 for Mus+musculus had 0.\n",
      "HINFP for Homo+sapiens had 2.\n",
      "MYBL1 for Homo+sapiens had 0.\n",
      "RAX for Homo+sapiens had 0.\n",
      "GRN for Drosophila+melanogaster had 0.\n",
      "SRY for Homo+sapiens had 1.\n",
      "NOTO for Homo+sapiens had 0.\n",
      "ZNF524 for Homo+sapiens had 0.\n",
      "NR2E1 for Mus+musculus had 0.\n",
      "POU3F3 for Homo+sapiens had 0.\n",
      "THRB for Homo+sapiens had 2.\n",
      "NKX3-1 for Mus+musculus had 0.\n",
      "UNCX for Homo+sapiens had 0.\n",
      "IRX3 for Mus+musculus had 0.\n",
      "ESRRA for Mus+musculus had 0.\n",
      "GLIS1 for Homo+sapiens had 1.\n",
      "UBX for Drosophila+melanogaster had 0.\n",
      "UNCX for Mus+musculus had 0.\n",
      "CRP for Drosophila+melanogaster had 0.\n",
      "DMRT99B for Drosophila+melanogaster had 0.\n",
      "KAY for Drosophila+melanogaster had 0.\n",
      "NAU for Drosophila+melanogaster had 0.\n",
      "CREBA for Drosophila+melanogaster had 0.\n",
      "HMX for Drosophila+melanogaster had 0.\n",
      "REPTOR-BP for Drosophila+melanogaster had 0.\n",
      "KAH for Drosophila+melanogaster had 0.\n",
      "CG34367 for Drosophila+melanogaster had 1.\n",
      "CWO for Drosophila+melanogaster had 0.\n",
      "KLF15 for Drosophila+melanogaster had 0.\n",
      "BAP for Drosophila+melanogaster had 0.\n",
      "SO for Drosophila+melanogaster had 0.\n",
      "E(SPL)M3-HLH for Drosophila+melanogaster had 0.\n",
      "CHES-1-LIKE for Drosophila+melanogaster had 0.\n",
      "VSX1 for Drosophila+melanogaster had 0.\n",
      "LIM3 for Drosophila+melanogaster had 0.\n",
      "FD102C for Drosophila+melanogaster had 0.\n",
      "HTH for Drosophila+melanogaster had 0.\n",
      "CTCF for Homo+sapiens had 457.\n",
      "EGR1 for Mus+musculus had 0.\n",
      "ALX4 for Mus+musculus had 0.\n",
      "ATOH1 for Mus+musculus had 0.\n",
      "BHLHE40 for Mus+musculus had 2.\n",
      "DLX2 for Mus+musculus had 0.\n",
      "E2F3 for Homo+sapiens had 1.\n",
      "ELF3 for Homo+sapiens had 1.\n",
      "ETS1 for Homo+sapiens had 6.\n",
      "ETV6 for Homo+sapiens had 7.\n",
      "FOXC1 for Mus+musculus had 0.\n",
      "FOXD3 for Homo+sapiens had 0.\n",
      "HNF1B for Homo+sapiens had 1.\n",
      "HNF4A for Homo+sapiens had 5.\n",
      "HOXD13 for Mus+musculus had 0.\n",
      "JDP2 for Homo+sapiens had 1.\n",
      "LHX2 for Homo+sapiens had 0.\n",
      "MAFG for Homo+sapiens had 2.\n",
      "NKX6-1 for Mus+musculus had 0.\n",
      "PRDM4 for Homo+sapiens had 2.\n",
      "RFX5 for Homo+sapiens had 9.\n",
      "RXRB for Mus+musculus had 0.\n",
      "SMAD3 for Homo+sapiens had 4.\n",
      "SP4 for Homo+sapiens had 2.\n",
      "SPI1 for Homo+sapiens had 4.\n",
      "TBX21 for Homo+sapiens had 1.\n",
      "THRA for Homo+sapiens had 2.\n",
      "XBP1 for Homo+sapiens had 1.\n",
      "ZBED1 for Homo+sapiens had 2.\n",
      "ZBTB49 for Homo+sapiens had 3.\n",
      "ZFP652 for Mus+musculus had 0.\n",
      "BTN for Drosophila+melanogaster had 0.\n",
      "DMRT93B for Drosophila+melanogaster had 0.\n",
      "DFD for Drosophila+melanogaster had 0.\n",
      "ASCL2 for Mus+musculus had 0.\n",
      "BHLHE40 for Homo+sapiens had 8.\n",
      "GBX1 for Homo+sapiens had 0.\n",
      "HMBOX1 for Mus+musculus had 0.\n",
      "HOXD3 for Mus+musculus had 0.\n",
      "MEF2A for Homo+sapiens had 4.\n",
      "NFIA for Homo+sapiens had 1.\n",
      "NKX2-8 for Homo+sapiens had 0.\n",
      "OTX1 for Mus+musculus had 0.\n",
      "SHOX2 for Mus+musculus had 0.\n",
      "TCF21 for Mus+musculus had 0.\n",
      "TEAD4 for Homo+sapiens had 11.\n",
      "TFAP2C for Homo+sapiens had 0.\n",
      "VAX1 for Homo+sapiens had 0.\n",
      "CAD for Drosophila+melanogaster had 0.\n",
      "PTX1 for Drosophila+melanogaster had 0.\n",
      "ARX for Mus+musculus had 0.\n",
      "CEBPB for Mus+musculus had 2.\n",
      "DLX1 for Mus+musculus had 0.\n",
      "EBF1 for Homo+sapiens had 2.\n",
      "EMX1 for Homo+sapiens had 1.\n",
      "ESRRA for Homo+sapiens had 8.\n",
      "FOXJ3 for Mus+musculus had 0.\n",
      "GBX1 for Mus+musculus had 0.\n",
      "GLIS2 for Homo+sapiens had 1.\n",
      "HLF for Mus+musculus had 0.\n",
      "HOXB5 for Homo+sapiens had 1.\n",
      "IRF4 for Homo+sapiens had 1.\n",
      "IRX2 for Homo+sapiens had 0.\n",
      "MEIS2 for Mus+musculus had 0.\n",
      "MSX1 for Homo+sapiens had 0.\n",
      "NR3C2 for Homo+sapiens had 0.\n",
      "PAX7 for Homo+sapiens had 0.\n",
      "PHOX2A for Homo+sapiens had 0.\n",
      "POU2F2 for Homo+sapiens had 2.\n",
      "POU2F3 for Homo+sapiens had 0.\n",
      "RHOX11 for Mus+musculus had 0.\n",
      "RUNX3 for Homo+sapiens had 1.\n",
      "SHOX2 for Homo+sapiens had 0.\n",
      "SOX1 for Mus+musculus had 0.\n",
      "SRF for Homo+sapiens had 12.\n",
      "TFAP4 for Homo+sapiens had 3.\n",
      "UNCX for Homo+sapiens had 0.\n",
      "USF1 for Homo+sapiens had 14.\n",
      "ZNF410 for Homo+sapiens had 0.\n",
      "HEY for Drosophila+melanogaster had 0.\n",
      "LAB for Drosophila+melanogaster had 0.\n",
      "SOX15 for Homo+sapiens had 1.\n",
      "RFX2 for Mus+musculus had 0.\n",
      "RXRA for Mus+musculus had 0.\n",
      "NKX6-1 for Homo+sapiens had 0.\n",
      "PRRX1 for Homo+sapiens had 0.\n",
      "UBX for Drosophila+melanogaster had 0.\n",
      "ABD-A for Drosophila+melanogaster had 0.\n",
      "ABD-B for Drosophila+melanogaster had 0.\n",
      "ANTP for Drosophila+melanogaster had 0.\n",
      "DFD for Drosophila+melanogaster had 0.\n",
      "LAB for Drosophila+melanogaster had 0.\n",
      "PB for Drosophila+melanogaster had 0.\n",
      "FOXP2 for Homo+sapiens had 2.\n",
      "FOSB for Homo+sapiens had 1.\n",
      "FOSL2 for Homo+sapiens had 19.\n",
      "FOSL2 for Homo+sapiens had 19.\n",
      "RARA for Homo+sapiens had 1.\n",
      "ZNF415 for Homo+sapiens had 0.\n",
      "ZNF682 for Homo+sapiens had 0.\n",
      "MINK1 for Homo+sapiens had 0.\n",
      "MINK1 for Homo+sapiens had 0.\n",
      "ALX1 for Mus+musculus had 0.\n",
      "CEBPD for Homo+sapiens had 1.\n",
      "CPEB1 for Homo+sapiens had 0.\n",
      "CUX2 for Homo+sapiens had 0.\n",
      "EGR3 for Mus+musculus had 0.\n",
      "ELK3 for Mus+musculus had 0.\n",
      "ESX1 for Homo+sapiens had 0.\n",
      "FEV for Homo+sapiens had 0.\n",
      "FOXD2 for Homo+sapiens had 0.\n",
      "FOXO6 for Homo+sapiens had 0.\n",
      "HIC2 for Homo+sapiens had 1.\n",
      "RNF31 for Homo+sapiens had 0.\n",
      "MEIS3 for Mus+musculus had 0.\n",
      "MEIS3P2 for Homo+sapiens had 0.\n",
      "MLXIPL for Homo+sapiens had 0.\n",
      "NEUROG2 for Homo+sapiens had 0.\n",
      "NFAT5 for Homo+sapiens had 1.\n",
      "NHLH1 for Homo+sapiens had 0.\n",
      "OLIG1 for Homo+sapiens had 0.\n",
      "OLIG2 for Homo+sapiens had 0.\n",
      "PAX1 for Homo+sapiens had 0.\n",
      "PAX2 for Homo+sapiens had 0.\n",
      "PDX1 for Homo+sapiens had 0.\n",
      "POU4F1 for Homo+sapiens had 0.\n",
      "POU4F2 for Homo+sapiens had 0.\n",
      "RAX2 for Homo+sapiens had 0.\n",
      "RORA for Homo+sapiens had 1.\n",
      "SPIC for Mus+musculus had 0.\n",
      "TBXT for Homo+sapiens had 0.\n",
      "TBX2 for Homo+sapiens had 1.\n",
      "TBX4 for Homo+sapiens had 0.\n",
      "TFCP2 for Homo+sapiens had 1.\n",
      "VDR for Homo+sapiens had 0.\n",
      "VDR for Mus+musculus had 0.\n",
      "VSX2 for Homo+sapiens had 0.\n",
      "ZBTB18 for Homo+sapiens had 0.\n",
      "ZNF713 for Homo+sapiens had 1.\n",
      "MEIS1 for Homo+sapiens had 2.\n",
      "BHLHE23 for Homo+sapiens had 0.\n",
      "GATAE for Drosophila+melanogaster had 0.\n",
      "CHRAC-14 for Drosophila+melanogaster had 0.\n",
      "CROC for Drosophila+melanogaster had 0.\n",
      "BS for Drosophila+melanogaster had 0.\n",
      "CG4956 for Drosophila+melanogaster had 0.\n",
      "SREBP for Drosophila+melanogaster had 0.\n",
      "BIGMAX for Drosophila+melanogaster had 0.\n",
      "ASH2 for Drosophila+melanogaster had 0.\n",
      "FTZ for Drosophila+melanogaster had 0.\n",
      "PNR for Drosophila+melanogaster had 0.\n",
      "POXM for Drosophila+melanogaster had 0.\n",
      "SRP for Drosophila+melanogaster had 0.\n",
      "FD96CA for Drosophila+melanogaster had 0.\n",
      "RX for Drosophila+melanogaster had 0.\n",
      "LBL for Drosophila+melanogaster had 0.\n",
      "DIMM for Drosophila+melanogaster had 0.\n",
      "HBN for Drosophila+melanogaster had 0.\n",
      "BIN for Drosophila+melanogaster had 0.\n",
      "TOY for Drosophila+melanogaster had 0.\n",
      "AOP for Drosophila+melanogaster had 0.\n",
      "POXN for Drosophila+melanogaster had 0.\n",
      "DOC2 for Drosophila+melanogaster had 0.\n",
      "ATF6 for Drosophila+melanogaster had 0.\n",
      "PDM2 for Drosophila+melanogaster had 0.\n",
      "MID for Drosophila+melanogaster had 0.\n",
      "BBX for Drosophila+melanogaster had 0.\n",
      "GT for Drosophila+melanogaster had 0.\n",
      "DIF for Drosophila+melanogaster had 0.\n",
      "CG14860 for Drosophila+melanogaster had 1.\n",
      "CBT for Drosophila+melanogaster had 0.\n",
      "SD for Drosophila+melanogaster had 0.\n",
      "KN for Drosophila+melanogaster had 0.\n",
      "CR4575 for Drosophila+melanogaster had 0.\n",
      "KLU for Drosophila+melanogaster had 0.\n",
      "GCM for Drosophila+melanogaster had 0.\n",
      "CG32532 for Drosophila+melanogaster had 1.\n",
      "ETS96B for Drosophila+melanogaster had 0.\n",
      "RFX for Drosophila+melanogaster had 0.\n",
      "THMG2 for Drosophila+melanogaster had 0.\n",
      "BYN for Drosophila+melanogaster had 0.\n",
      "ETS65A for Drosophila+melanogaster had 0.\n",
      "CG12605 for Drosophila+melanogaster had 1.\n",
      "LMD for Drosophila+melanogaster had 0.\n",
      "ABD-A for Drosophila+melanogaster had 0.\n",
      "CG11085 for Drosophila+melanogaster had 1.\n",
      "VVL for Drosophila+melanogaster had 0.\n",
      "SUG for Drosophila+melanogaster had 0.\n",
      "LMX1A for Drosophila+melanogaster had 0.\n",
      "AL for Drosophila+melanogaster had 0.\n",
      "TWI for Drosophila+melanogaster had 0.\n",
      "ATF4 for Homo+sapiens had 5.\n",
      "ATF4 for Homo+sapiens had 5.\n",
      "CEBPG for Homo+sapiens had 5.\n",
      "BATF3 for Homo+sapiens had 0.\n",
      "FOS for Homo+sapiens had 12.\n",
      "FOS for Homo+sapiens had 12.\n",
      "JUNB for Homo+sapiens had 18.\n",
      "JUNB for Homo+sapiens had 18.\n",
      "JUNB for Homo+sapiens had 18.\n",
      "JUNB for Homo+sapiens had 18.\n",
      "HOXA1 for Homo+sapiens had 0.\n",
      "HOXA9 for Homo+sapiens had 1.\n",
      "EXD for Drosophila+melanogaster had 0.\n",
      "SCR for Drosophila+melanogaster had 0.\n",
      "SD for Drosophila+melanogaster had 0.\n",
      "FOS for Homo+sapiens had 12.\n",
      "OPA for Drosophila+melanogaster had 0.\n",
      "PPARG for Mus+musculus had 0.\n",
      "ZFP14 for Homo+sapiens had 1.\n",
      "CENPB for Homo+sapiens had 0.\n",
      "HEY1 for Homo+sapiens had 1.\n",
      "ZIC4 for Homo+sapiens had 0.\n",
      "TRP73 for Mus+musculus had 0.\n",
      "HGTX for Drosophila+melanogaster had 0.\n",
      "SOX21A for Drosophila+melanogaster had 0.\n",
      "OTP for Drosophila+melanogaster had 0.\n",
      "WOR for Drosophila+melanogaster had 0.\n",
      "LMS for Drosophila+melanogaster had 0.\n",
      "PPH13 for Drosophila+melanogaster had 0.\n",
      "SEQ for Drosophila+melanogaster had 0.\n",
      "GEM for Drosophila+melanogaster had 0.\n",
      "RN for Drosophila+melanogaster had 0.\n",
      "CEBPA for Homo+sapiens had 1.\n",
      "ATF5 for Homo+sapiens had 1.\n",
      "JUNB for Homo+sapiens had 18.\n",
      "CEBPA for Homo+sapiens had 1.\n",
      "CEBPG for Homo+sapiens had 5.\n",
      "ATF4 for Homo+sapiens had 5.\n",
      "BATF3 for Homo+sapiens had 0.\n",
      "FOS for Homo+sapiens had 12.\n",
      "FOS for Homo+sapiens had 12.\n",
      "FOS for Homo+sapiens had 12.\n",
      "FOS for Homo+sapiens had 12.\n",
      "FOS for Homo+sapiens had 12.\n",
      "FOSL1 for Homo+sapiens had 6.\n",
      "ATF3 for Homo+sapiens had 14.\n",
      "CEBPA for Homo+sapiens had 1.\n",
      "JUNB for Homo+sapiens had 18.\n",
      "JUNB for Homo+sapiens had 18.\n",
      "JUNB for Homo+sapiens had 18.\n",
      "JUNB for Homo+sapiens had 18.\n",
      "JUNB for Homo+sapiens had 18.\n",
      "FOS for Homo+sapiens had 12.\n",
      "FOSL1 for Homo+sapiens had 6.\n",
      "FOSL1 for Homo+sapiens had 6.\n",
      "PBX1 for Homo+sapiens had 1.\n",
      "MINK1 for Homo+sapiens had 0.\n",
      "AR for Homo+sapiens had 1.\n",
      "AR for Mus+musculus had 0.\n",
      "CREB5 for Mus+musculus had 0.\n",
      "DRGX for Homo+sapiens had 0.\n",
      "DUXA for Homo+sapiens had 0.\n",
      "ELF5 for Homo+sapiens had 0.\n",
      "ELF5 for Mus+musculus had 0.\n",
      "ERG for Homo+sapiens had 1.\n",
      "ESRRG for Homo+sapiens had 1.\n",
      "FOXC2 for Homo+sapiens had 0.\n",
      "FOXO1 for Homo+sapiens had 1.\n",
      "FOXO3 for Homo+sapiens had 0.\n",
      "FOXO4 for Homo+sapiens had 2.\n",
      "GRHL1 for Homo+sapiens had 1.\n",
      "HESX1 for Homo+sapiens had 0.\n",
      "HOXD9 for Mus+musculus had 0.\n",
      "KLF12 for Mus+musculus had 0.\n",
      "MEOX2 for Homo+sapiens had 0.\n",
      "MEOX2 for Mus+musculus had 0.\n",
      "NFIX for Homo+sapiens had 1.\n",
      "ONECUT3 for Homo+sapiens had 0.\n",
      "PAX3 for Homo+sapiens had 0.\n",
      "POU6F2 for Homo+sapiens had 0.\n",
      "SOX3 for Mus+musculus had 0.\n",
      "SOX9 for Homo+sapiens had 0.\n",
      "TBR1 for Mus+musculus had 0.\n",
      "TBX1 for Homo+sapiens had 0.\n",
      "TBX15 for Homo+sapiens had 0.\n",
      "TBX19 for Homo+sapiens had 0.\n",
      "TBX20 for Homo+sapiens had 0.\n",
      "TBX5 for Homo+sapiens had 0.\n",
      "ZBTB7C for Homo+sapiens had 0.\n",
      "ZKSCAN3 for Homo+sapiens had 1.\n",
      "AC020909.1 for Homo+sapiens had 0.\n",
      "NFKB2 for Homo+sapiens had 1.\n",
      "DOC3 for Drosophila+melanogaster had 0.\n",
      "HLH4C for Drosophila+melanogaster had 1.\n",
      "GSB for Drosophila+melanogaster had 0.\n",
      "TFAP-2 for Drosophila+melanogaster had 0.\n",
      "CG14667 for Drosophila+melanogaster had 1.\n",
      "CG11294 for Drosophila+melanogaster had 1.\n",
      "ONECUT for Drosophila+melanogaster had 0.\n",
      "BOWL for Drosophila+melanogaster had 0.\n",
      "SLBO for Drosophila+melanogaster had 0.\n",
      "E5 for Drosophila+melanogaster had 0.\n",
      "DRGX for Drosophila+melanogaster had 0.\n",
      "NUB for Drosophila+melanogaster had 0.\n",
      "CG4360 for Drosophila+melanogaster had 1.\n",
      "AEF1 for Drosophila+melanogaster had 0.\n",
      "SNA for Drosophila+melanogaster had 0.\n",
      "DATI for Drosophila+melanogaster had 0.\n",
      "ATF3 for Homo+sapiens had 14.\n",
      "CEBPG for Homo+sapiens had 5.\n",
      "CEBPE for Homo+sapiens had 0.\n",
      "FOS for Homo+sapiens had 12.\n",
      "ARX for Homo+sapiens had 0.\n",
      "BARHL2 for Homo+sapiens had 0.\n",
      "BATF3 for Homo+sapiens had 0.\n",
      "BHLHE41 for Homo+sapiens had 0.\n",
      "CDX2 for Homo+sapiens had 0.\n",
      "CLOCK for Homo+sapiens had 3.\n",
      "DLX4 for Homo+sapiens had 1.\n",
      "DLX5 for Homo+sapiens had 0.\n",
      "DLX6 for Homo+sapiens had 1.\n",
      "DPRX for Homo+sapiens had 0.\n",
      "E2F4 for Homo+sapiens had 8.\n",
      "E2F7 for Homo+sapiens had 1.\n",
      "E2F8 for Homo+sapiens had 4.\n",
      "EGR4 for Homo+sapiens had 0.\n",
      "ELK3 for Homo+sapiens had 1.\n",
      "EN1 for Homo+sapiens had 0.\n",
      "ETV4 for Homo+sapiens had 2.\n",
      "FOXB1 for Homo+sapiens had 0.\n",
      "FOXJ2 for Homo+sapiens had 1.\n",
      "GATA4 for Homo+sapiens had 1.\n",
      "GATA5 for Homo+sapiens had 0.\n",
      "GCM2 for Homo+sapiens had 0.\n",
      "GLI2 for Homo+sapiens had 1.\n",
      "GSX1 for Homo+sapiens had 0.\n",
      "HOXA1 for Homo+sapiens had 0.\n",
      "HOXB2 for Homo+sapiens had 0.\n",
      "HOXC10 for Homo+sapiens had 1.\n",
      "HOXC11 for Homo+sapiens had 0.\n",
      "HOXC12 for Homo+sapiens had 0.\n",
      "HOXD11 for Homo+sapiens had 0.\n",
      "HOXD12 for Homo+sapiens had 0.\n",
      "HOXD8 for Homo+sapiens had 0.\n",
      "HSF1 for Homo+sapiens had 3.\n",
      "HSF4 for Homo+sapiens had 1.\n",
      "HSFY2 for Homo+sapiens had 0.\n",
      "IRF3 for Homo+sapiens had 5.\n",
      "IRF5 for Homo+sapiens had 2.\n",
      "IRF8 for Homo+sapiens had 0.\n",
      "IRX5 for Homo+sapiens had 0.\n",
      "KLF14 for Homo+sapiens had 0.\n",
      "KLF16 for Homo+sapiens had 4.\n",
      "LMX1B for Homo+sapiens had 0.\n",
      "MEF2B for Homo+sapiens had 1.\n",
      "MEIS2 for Homo+sapiens had 2.\n",
      "MIXL1 for Homo+sapiens had 1.\n",
      "MSC for Homo+sapiens had 0.\n",
      "MYF6 for Homo+sapiens had 0.\n",
      "NKX2-3 for Homo+sapiens had 0.\n",
      "NKX6-2 for Homo+sapiens had 0.\n",
      "OLIG3 for Homo+sapiens had 0.\n",
      "ONECUT2 for Homo+sapiens had 1.\n",
      "POU1F1 for Homo+sapiens had 0.\n",
      "PROP1 for Homo+sapiens had 0.\n",
      "PRRX2 for Homo+sapiens had 1.\n",
      "RFX3 for Homo+sapiens had 1.\n",
      "RHOXF1 for Homo+sapiens had 0.\n",
      "RUNX2 for Homo+sapiens had 0.\n",
      "SOX10 for Homo+sapiens had 0.\n",
      "SOX4 for Homo+sapiens had 0.\n",
      "SOX8 for Homo+sapiens had 0.\n",
      "SP8 for Homo+sapiens had 0.\n",
      "TFE3 for Homo+sapiens had 2.\n",
      "TGIF2 for Homo+sapiens had 3.\n",
      "TGIF2LX for Homo+sapiens had 0.\n",
      "YY2 for Homo+sapiens had 1.\n",
      "ZBTB7A for Homo+sapiens had 5.\n",
      "ZBTB7B for Homo+sapiens had 2.\n",
      "ZNF784 for Homo+sapiens had 1.\n",
      "ZSCAN4 for Homo+sapiens had 1.\n",
      "ELF1 for Homo+sapiens had 14.\n",
      "NRL for Homo+sapiens had 1.\n",
      "ESRRB for Homo+sapiens had 1.\n",
      "FIGLA for Homo+sapiens had 0.\n",
      "HES7 for Homo+sapiens had 0.\n",
      "HOXD13 for Homo+sapiens had 1.\n",
      "NR1I3 for Homo+sapiens had 0.\n",
      "ARGFX for Homo+sapiens had 0.\n",
      "ARNT2 for Homo+sapiens had 1.\n",
      "ASCL1 for Homo+sapiens had 0.\n",
      "ASCL2 for Homo+sapiens had 0.\n",
      "ATF4 for Homo+sapiens had 5.\n",
      "ZNF597 for Homo+sapiens had 1.\n",
      "ZNF660 for Homo+sapiens had 2.\n",
      "ZNF684 for Homo+sapiens had 0.\n",
      "ZNF704 for Homo+sapiens had 2.\n",
      "ATF6 for Homo+sapiens had 3.\n",
      "ATF6B for Homo+sapiens had 0.\n",
      "ATOH1 for Homo+sapiens had 0.\n",
      "BACH2 for Homo+sapiens had 1.\n",
      "BBX for Homo+sapiens had 0.\n",
      "CREM for Homo+sapiens had 5.\n",
      "DMRT1 for Homo+sapiens had 0.\n",
      "DMRTA1 for Homo+sapiens had 0.\n",
      "DMRTA2 for Homo+sapiens had 0.\n",
      "DMRTC2 for Homo+sapiens had 0.\n",
      "DPF1 for Homo+sapiens had 0.\n",
      "ESR1 for Homo+sapiens had 9.\n",
      "ETS2 for Homo+sapiens had 1.\n",
      "ETV2 for Homo+sapiens had 0.\n",
      "ETV3 for Homo+sapiens had 0.\n",
      "ETV7 for Homo+sapiens had 0.\n",
      "FERD3L for Homo+sapiens had 0.\n",
      "FOSB for Homo+sapiens had 1.\n",
      "FOXA1 for Homo+sapiens had 13.\n",
      "FOXA2 for Homo+sapiens had 6.\n",
      "FOXA3 for Homo+sapiens had 3.\n",
      "FOXE1 for Homo+sapiens had 0.\n",
      "FOXP1 for Homo+sapiens had 4.\n",
      "GATA2 for Homo+sapiens had 8.\n",
      "GATA6 for Homo+sapiens had 0.\n",
      "GFI1B for Homo+sapiens had 1.\n",
      "HAND2 for Homo+sapiens had 0.\n",
      "HES2 for Homo+sapiens had 12.\n",
      "HES6 for Homo+sapiens had 0.\n",
      "HIC1 for Homo+sapiens had 1.\n",
      "HMBOX1 for Homo+sapiens had 1.\n",
      "HOXA4 for Homo+sapiens had 0.\n",
      "HOXA5 for Homo+sapiens had 1.\n",
      "HOXA7 for Homo+sapiens had 3.\n",
      "HOXA9 for Homo+sapiens had 1.\n",
      "HOXB1 for Homo+sapiens had 0.\n",
      "HOXB4 for Homo+sapiens had 0.\n",
      "HOXB7 for Homo+sapiens had 1.\n",
      "HOXB9 for Homo+sapiens had 0.\n",
      "HOXC8 for Homo+sapiens had 0.\n",
      "HOXC9 for Homo+sapiens had 0.\n",
      "HOXD10 for Homo+sapiens had 0.\n",
      "HOXD4 for Homo+sapiens had 0.\n",
      "HOXD9 for Homo+sapiens had 0.\n",
      "IRF2 for Homo+sapiens had 4.\n",
      "IRF6 for Homo+sapiens had 0.\n",
      "IRX1 for Homo+sapiens had 0.\n",
      "KLF10 for Homo+sapiens had 2.\n",
      "KLF11 for Homo+sapiens had 1.\n",
      "KLF12 for Homo+sapiens had 1.\n",
      "KLF15 for Homo+sapiens had 1.\n",
      "KLF2 for Homo+sapiens had 0.\n",
      "KLF4 for Homo+sapiens had 1.\n",
      "KLF5 for Homo+sapiens had 1.\n",
      "LHX1 for Homo+sapiens had 0.\n",
      "LHX5 for Homo+sapiens had 0.\n",
      "MAF for Homo+sapiens had 0.\n",
      "MAFA for Homo+sapiens had 0.\n",
      "MEIS3 for Homo+sapiens had 0.\n",
      "MSGN1 for Homo+sapiens had 0.\n",
      "MYOD1 for Homo+sapiens had 0.\n",
      "MYOG for Homo+sapiens had 0.\n",
      "NEUROD1 for Homo+sapiens had 2.\n",
      "NEUROG1 for Homo+sapiens had 0.\n",
      "NFATC3 for Homo+sapiens had 4.\n",
      "NHLH2 for Homo+sapiens had 0.\n",
      "NKX6-3 for Homo+sapiens had 0.\n",
      "NPAS2 for Homo+sapiens had 2.\n",
      "NR1D1 for Homo+sapiens had 0.\n",
      "NR1D2 for Homo+sapiens had 0.\n",
      "NR2C1 for Homo+sapiens had 3.\n",
      "NR5A1 for Homo+sapiens had 1.\n",
      "NR5A2 for Homo+sapiens had 1.\n",
      "NR6A1 for Homo+sapiens had 0.\n",
      "OSR1 for Homo+sapiens had 1.\n",
      "OSR2 for Homo+sapiens had 1.\n",
      "OVOL1 for Homo+sapiens had 1.\n",
      "OVOL2 for Homo+sapiens had 1.\n",
      "PAX8 for Homo+sapiens had 3.\n",
      "POU6F1 for Homo+sapiens had 1.\n",
      "PPARD for Homo+sapiens had 1.\n",
      "RFX7 for Homo+sapiens had 1.\n",
      "RHOXF2 for Homo+sapiens had 0.\n",
      "RORB for Homo+sapiens had 1.\n",
      "SKOR1 for Homo+sapiens had 0.\n",
      "SKOR2 for Homo+sapiens had 0.\n",
      "SOX3 for Homo+sapiens had 0.\n",
      "SP9 for Homo+sapiens had 0.\n",
      "SPIB for Homo+sapiens had 0.\n",
      "TBX3 for Homo+sapiens had 2.\n",
      "TBX6 for Homo+sapiens had 0.\n",
      "TCF21 for Homo+sapiens had 0.\n",
      "TEAD2 for Homo+sapiens had 2.\n",
      "TFAP2E for Homo+sapiens had 0.\n",
      "TGIF2LY for Homo+sapiens had 0.\n",
      "TLX2 for Homo+sapiens had 0.\n",
      "UBP1 for Homo+sapiens had 0.\n",
      "USF2 for Homo+sapiens had 12.\n",
      "ZBTB12 for Homo+sapiens had 2.\n",
      "ZBTB2 for Homo+sapiens had 2.\n",
      "ZBTB20 for Homo+sapiens had 2.\n",
      "ZBTB45 for Homo+sapiens had 0.\n",
      "ZFP41 for Homo+sapiens had 2.\n",
      "ZIC5 for Homo+sapiens had 0.\n",
      "ZNF140 for Homo+sapiens had 2.\n",
      "ZNF174 for Homo+sapiens had 2.\n",
      "ZNF276 for Homo+sapiens had 1.\n",
      "ZNF281 for Homo+sapiens had 3.\n",
      "ZNF296 for Homo+sapiens had 1.\n",
      "ZNF32 for Homo+sapiens had 0.\n",
      "ZNF323 for Homo+sapiens had 0.\n",
      "ZNF343 for Homo+sapiens had 1.\n",
      "ZNF385D for Homo+sapiens had 0.\n",
      "ZNF396 for Homo+sapiens had 1.\n",
      "ZNF444 for Homo+sapiens had 2.\n",
      "ZNF771 for Homo+sapiens had 0.\n",
      "ZNF821 for Homo+sapiens had 0.\n",
      "ZSCAN1 for Homo+sapiens had 0.\n",
      "ZSCAN5 for Homo+sapiens had 0.\n",
      "ZSCAN9 for Homo+sapiens had 1.\n",
      "EGR2 for Homo+sapiens had 1.\n",
      "FOXI1 for Homo+sapiens had 0.\n",
      "GSC2 for Homo+sapiens had 0.\n",
      "HMX2 for Homo+sapiens had 0.\n",
      "HOXB13 for Homo+sapiens had 1.\n",
      "HOXC13 for Homo+sapiens had 0.\n",
      "KLF13 for Homo+sapiens had 2.\n",
      "MSX2 for Homo+sapiens had 1.\n",
      "NFIB for Homo+sapiens had 3.\n",
      "NR4A2 for Homo+sapiens had 0.\n",
      "SPDEF for Homo+sapiens had 1.\n",
      "TFEC for Homo+sapiens had 0.\n",
      "ZIC1 for Homo+sapiens had 0.\n",
      "ZNF713 for Homo+sapiens had 1.\n",
      "RFX4 for Homo+sapiens had 0.\n",
      "BARHL1 for Homo+sapiens had 0.\n",
      "BARX2 for Homo+sapiens had 0.\n",
      "BCL6 for Homo+sapiens had 2.\n",
      "ELF4 for Homo+sapiens had 6.\n",
      "FOXR2 for Homo+sapiens had 0.\n",
      "GLI3 for Homo+sapiens had 0.\n",
      "GMEB1 for Homo+sapiens had 3.\n",
      "HOXD1 for Homo+sapiens had 1.\n",
      "HOXD3 for Homo+sapiens had 0.\n",
      "HSF5 for Homo+sapiens had 0.\n",
      "NKX2-5 for Homo+sapiens had 0.\n",
      "PITX2 for Homo+sapiens had 0.\n",
      "RFX1 for Homo+sapiens had 5.\n",
      "SOX30 for Homo+sapiens had 0.\n",
      "TCF12 for Homo+sapiens had 12.\n",
      "TFCP2L1 for Homo+sapiens had 1.\n",
      "ARNTL for Homo+sapiens had 2.\n",
      "ATF7 for Homo+sapiens had 5.\n",
      "BCL6B for Homo+sapiens had 1.\n",
      "ALX1 for Homo+sapiens had 0.\n",
      "CDX1 for Homo+sapiens had 0.\n",
      "CREB3 for Homo+sapiens had 2.\n",
      "DLX3 for Homo+sapiens had 0.\n",
      "DMBX1 for Homo+sapiens had 1.\n",
      "EGR3 for Homo+sapiens had 0.\n",
      "GATA3 for Homo+sapiens had 8.\n",
      "GSC for Homo+sapiens had 0.\n",
      "GSX2 for Homo+sapiens had 0.\n",
      "HES5 for Homo+sapiens had 0.\n",
      "HOXA10 for Pantherophis+guttatus had 0.\n",
      "HOXA2 for Homo+sapiens had 0.\n",
      "HSF2 for Homo+sapiens had 1.\n",
      "ISL2 for Homo+sapiens had 1.\n",
      "LBX2 for Homo+sapiens had 1.\n",
      "LHX9 for Homo+sapiens had 0.\n",
      "MAFF for Homo+sapiens had 3.\n",
      "MEOX1 for Homo+sapiens had 0.\n",
      "MTF1 for Homo+sapiens had 2.\n",
      "NFATC1 for Homo+sapiens had 1.\n",
      "NR2E1 for Homo+sapiens had 0.\n",
      "NR2F1 for Homo+sapiens had 4.\n",
      "ONECUT1 for Homo+sapiens had 1.\n",
      "PAX9 for Homo+sapiens had 0.\n",
      "PHOX2B for Homo+sapiens had 0.\n",
      "PITX3 for Homo+sapiens had 0.\n",
      "POU4F3 for Homo+sapiens had 0.\n",
      "RARB for Homo+sapiens had 2.\n",
      "SOX7 for Homo+sapiens had 0.\n",
      "TCF7L1 for Homo+sapiens had 0.\n",
      "ZSCAN16 for Homo+sapiens had 1.\n",
      "ZNF75A for Homo+sapiens had 1.\n",
      "ELK4 for Homo+sapiens had 3.\n",
      "SNAI2 for Homo+sapiens had 1.\n",
      "FOXK1 for Homo+sapiens had 3.\n",
      "ATF3 for Homo+sapiens had 14.\n",
      "CDX4 for Homo+sapiens had 1.\n",
      "CREB3L4 for Homo+sapiens had 0.\n",
      "CREB5 for Homo+sapiens had 2.\n",
      "CRX for Homo+sapiens had 0.\n",
      "DMRT3 for Homo+sapiens had 0.\n",
      "ELF2 for Homo+sapiens had 1.\n",
      "FOXP3 for Homo+sapiens had 0.\n",
      "GFI1 for Homo+sapiens had 1.\n",
      "HES1 for Homo+sapiens had 2.\n",
      "HOXA11 for Homo+sapiens had 0.\n",
      "HOXA6 for Homo+sapiens had 0.\n",
      "HOXB8 for Homo+sapiens had 0.\n",
      "HOXC4 for Homo+sapiens had 0.\n",
      "IRF9 for Homo+sapiens had 2.\n",
      "IRX3 for Homo+sapiens had 1.\n",
      "ISL1 for Homo+sapiens had 1.\n",
      "KLF3 for Homo+sapiens had 0.\n",
      "LHX8 for Homo+sapiens had 0.\n",
      "NFATC2 for Homo+sapiens had 0.\n",
      "NFATC4 for Homo+sapiens had 1.\n",
      "NFIC for Homo+sapiens had 5.\n",
      "NR4A1 for Homo+sapiens had 3.\n",
      "POU5F1 for Homo+sapiens had 4.\n",
      "RORC for Homo+sapiens had 0.\n",
      "SIX1 for Homo+sapiens had 1.\n",
      "SIX2 for Homo+sapiens had 0.\n",
      "SIX3 for Homo+sapiens had 0.\n",
      "SNAI1 for Homo+sapiens had 1.\n",
      "SNAI3 for Homo+sapiens had 0.\n",
      "SOX12 for Homo+sapiens had 0.\n",
      "YBX1 for Homo+sapiens had 4.\n",
      "ZBTB26 for Homo+sapiens had 3.\n",
      "ZNF274 for Homo+sapiens had 12.\n",
      "ZSCAN23 for Homo+sapiens had 1.\n",
      "FLI1 for Homo+sapiens had 0.\n",
      "GCM1 for Homo+sapiens had 0.\n",
      "LEF1 for Homo+sapiens had 3.\n",
      "LHX6 for Homo+sapiens had 0.\n",
      "CEBPE for Homo+sapiens had 0.\n",
      "HLF for Homo+sapiens had 1.\n",
      "HMX1 for Homo+sapiens had 0.\n",
      "MLX for Homo+sapiens had 4.\n",
      "PKNOX2 for Homo+sapiens had 0.\n",
      "RXRB for Homo+sapiens had 1.\n",
      "TCF4 for Homo+sapiens had 1.\n",
      "TFAP2A for Homo+sapiens had 0.\n",
      "TGIF1 for Homo+sapiens had 0.\n",
      "CEBPB for Homo+sapiens had 28.\n",
      "YY1 for Homo+sapiens had 18.\n",
      "BSX for Homo+sapiens had 0.\n",
      "ETV5 for Homo+sapiens had 2.\n",
      "GMEB2 for Homo+sapiens had 1.\n",
      "HOXA13 for Homo+sapiens had 0.\n",
      "PRDM1 for Homo+sapiens had 3.\n",
      "SCRT1 for Homo+sapiens had 1.\n",
      "ZIC3 for Homo+sapiens had 0.\n",
      "SP3 for Homo+sapiens had 1.\n",
      "CREB1 for Homo+sapiens had 15.\n",
      "GABPA for Homo+sapiens had 14.\n",
      "RARA for Homo+sapiens had 1.\n",
      "BHLHA15 for Homo+sapiens had 1.\n",
      "E2F2 for Homo+sapiens had 1.\n",
      "ELK1 for Homo+sapiens had 8.\n",
      "EVX2 for Homo+sapiens had 0.\n",
      "FOXJ3 for Homo+sapiens had 4.\n",
      "MNX1 for Homo+sapiens had 1.\n",
      "MYBL2 for Homo+sapiens had 5.\n",
      "NR3C1 for Homo+sapiens had 32.\n",
      "OTX1 for Homo+sapiens had 2.\n",
      "RFX2 for Homo+sapiens had 0.\n",
      "SCRT2 for Homo+sapiens had 1.\n",
      "SHOX for Homo+sapiens had 0.\n",
      "TEAD1 for Homo+sapiens had 4.\n",
      "ATF2 for Homo+sapiens had 9.\n"
     ]
    }
   ],
   "source": [
    "hits = {}\n",
    "have_hits = 0\n",
    "\n",
    "for id, gene, taxa in  [fit_info(fit) for fit in MOTIFCENTRAL]:\n",
    "   search_results = ENCODE_search(gene.upper(), tax_id_dict[taxa])\n",
    "   hits[id] = len(search_results[\"@graph\"])\n",
    "   if hits[id] >0: have_hits+=1\n",
    "   print(f'{gene.upper()} for {tax_id_dict[taxa]} had {hits[id]}.')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "337"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "have_hits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.38558352402745993"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "have_hits/len(MOTIFCENTRAL)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_values([0, 0, 14, 7, 0, 0, 0, 0, 1, 0, 0, 0, 23, 5, 1, 1, 0, 0, 1, 0, 0, 0, 5, 2, 4, 0, 0, 0, 0, 0, 0, 0, 4, 0, 0, 0, 0, 1, 0, 7, 9, 5, 0, 0, 1, 2, 0, 0, 0, 0, 0, 3, 1, 0, 0, 8, 0, 0, 0, 0, 2, 0, 8, 2, 0, 0, 1, 2, 1, 0, 1, 0, 0, 0, 1, 0, 3, 0, 0, 0, 0, 0, 4, 4, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 15, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 12, 1, 1, 4, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 3, 1, 0, 0, 0, 0, 0, 1, 0, 0, 2, 0, 0, 0, 1, 0, 0, 0, 0, 2, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 457, 0, 0, 0, 2, 0, 1, 1, 6, 7, 0, 0, 1, 5, 0, 1, 0, 2, 0, 2, 9, 0, 4, 2, 4, 1, 2, 1, 2, 3, 0, 0, 0, 0, 0, 8, 0, 0, 0, 4, 1, 0, 0, 0, 0, 11, 0, 0, 0, 0, 0, 2, 0, 2, 1, 8, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 2, 0, 0, 1, 0, 0, 12, 3, 0, 14, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 1, 19, 19, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 5, 5, 5, 0, 12, 12, 18, 18, 18, 18, 0, 1, 0, 0, 0, 12, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 18, 1, 5, 5, 0, 12, 12, 12, 12, 12, 6, 14, 1, 18, 18, 18, 18, 18, 12, 6, 6, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 2, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 14, 5, 0, 12, 0, 0, 0, 0, 0, 3, 1, 0, 1, 0, 8, 1, 4, 0, 1, 0, 2, 0, 1, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 3, 1, 0, 5, 2, 0, 0, 0, 4, 0, 1, 2, 1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 2, 3, 0, 1, 5, 2, 1, 1, 14, 1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 5, 1, 2, 0, 2, 3, 0, 0, 1, 0, 5, 0, 0, 0, 0, 0, 9, 1, 0, 0, 0, 0, 1, 13, 6, 3, 0, 4, 8, 0, 1, 0, 12, 0, 1, 1, 0, 1, 3, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 4, 0, 0, 2, 1, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 2, 0, 4, 0, 0, 2, 0, 0, 3, 1, 1, 0, 1, 1, 1, 1, 3, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 2, 0, 0, 2, 0, 0, 0, 0, 12, 2, 2, 2, 0, 2, 0, 2, 2, 1, 3, 1, 0, 0, 1, 0, 1, 2, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 2, 1, 3, 0, 1, 0, 0, 1, 0, 0, 0, 2, 6, 0, 0, 3, 1, 0, 0, 0, 0, 5, 0, 12, 1, 2, 5, 1, 0, 0, 2, 0, 1, 0, 8, 0, 0, 0, 0, 0, 1, 1, 1, 0, 3, 0, 2, 1, 0, 4, 1, 0, 0, 0, 0, 2, 0, 0, 1, 1, 3, 1, 3, 14, 1, 0, 2, 0, 0, 1, 0, 1, 2, 0, 0, 0, 0, 2, 1, 1, 0, 0, 0, 1, 5, 3, 4, 0, 1, 0, 0, 1, 0, 0, 4, 3, 12, 1, 0, 0, 3, 0, 0, 1, 0, 4, 0, 1, 1, 0, 0, 28, 18, 0, 2, 1, 0, 3, 1, 0, 1, 15, 14, 1, 1, 1, 8, 0, 4, 1, 5, 32, 2, 0, 1, 0, 4, 9])"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hits.values()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([24.,  3.,  8.,  4.,  0.,  1., 16.,  1.,  7.,  2.]),\n",
       " array([ 5. ,  6.1,  7.2,  8.3,  9.4, 10.5, 11.6, 12.7, 13.8, 14.9, 16. ]),\n",
       " <BarContainer object of 10 artists>)"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAh8AAAGfCAYAAAD/BbCUAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy81sbWrAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAZbUlEQVR4nO3da4wV9d3A8d8pyAHssgYpe6mAmKK2QtCIwSLl0jxSqcELWi9Ehdg2GlGL1AuUGrdtBKQRaUrEYBqqsVTfINXSilgFNFRFELVoFeuia3VDtLoLiOtl53nRuOkKCgfP/s8e/HySSZw5s2d+Thb2mzmzTC7LsiwAABL5SqkHAAC+XMQHAJCU+AAAkhIfAEBS4gMASEp8AABJiQ8AICnxAQAkJT4AgKTEBwCQVNdCdp4zZ04sW7Ys/vnPf0aPHj1ixIgRcdNNN8VRRx3Vts+UKVPijjvuaPd1w4cPj8cff3yfjtHa2hpvvPFGVFRURC6XK2Q8AKBEsiyL7du3R21tbXzlK59/baOg+FizZk1MnTo1TjjhhPjoo49i1qxZMW7cuHj++efj4IMPbtvvlFNOiSVLlrStd+vWbZ+P8cYbb0S/fv0KGQsA6CQaGhrisMMO+9x9CoqPBx54oN36kiVLom/fvrFhw4YYNWpU2/Z8Ph/V1dWFvHWbioqKiPjv8L169dqv9wAA0mpubo5+/fq1/Rz/PAXFx6c1NTVFRETv3r3bbV+9enX07ds3DjnkkBg9enTceOON0bdv3z2+R0tLS7S0tLStb9++PSIievXqJT4AoMzsyy0TuSzLsv158yzL4vTTT4933nknHn300bbt99xzT3z1q1+NAQMGRH19fVx//fXx0UcfxYYNGyKfz+/2PnV1dfGLX/xit+1NTU3iAwDKRHNzc1RWVu7Tz+/9jo+pU6fGihUr4rHHHvvcz3befPPNGDBgQNx9990xceLE3V7/9JWPTy7biA8AKB+FxMd+fexyxRVXxH333Rdr167d600lNTU1MWDAgNiyZcseX8/n83u8IgIAHJgKio8sy+KKK66Ie++9N1avXh0DBw7c69e8/fbb0dDQEDU1Nfs9JABw4CjoHxmbOnVq3HXXXbF06dKoqKiIxsbGaGxsjF27dkVExI4dO+Lqq6+Ov//977F169ZYvXp1TJgwIfr06RNnnnlmh/wPAADlpaB7Pj7rDtYlS5bElClTYteuXXHGGWfE008/He+++27U1NTE2LFj41e/+tU+/9sdhXxmBAB0Dh12z8feOqVHjx6xcuXKQt4SAPiS8WwXACAp8QEAJCU+AICkxAcAkJT4AACSEh8AQFLiAwBIar+e7VLODp+xotQjFGzr3FNLPQIAFI0rHwBAUuIDAEhKfAAASYkPACAp8QEAJCU+AICkxAcAkJT4AACSEh8AQFLiAwBISnwAAEmJDwAgKfEBACQlPgCApMQHAJCU+AAAkhIfAEBS4gMASEp8AABJiQ8AICnxAQAkJT4AgKTEBwCQlPgAAJISHwBAUuIDAEhKfAAASYkPACAp8QEAJCU+AICkxAcAkJT4AACSEh8AQFLiAwBISnwAAEmJDwAgKfEBACQlPgCApMQHAJCU+AAAkhIfAEBS4gMASEp8AABJiQ8AICnxAQAkJT4AgKTEBwCQlPgAAJISHwBAUuIDAEhKfAAASYkPACAp8QEAJCU+AICkCoqPOXPmxAknnBAVFRXRt2/fOOOMM+LFF19st0+WZVFXVxe1tbXRo0ePGDNmTGzevLmoQwMA5aug+FizZk1MnTo1Hn/88Vi1alV89NFHMW7cuNi5c2fbPvPmzYv58+fHwoULY/369VFdXR0nn3xybN++vejDAwDlp2shOz/wwAPt1pcsWRJ9+/aNDRs2xKhRoyLLsliwYEHMmjUrJk6cGBERd9xxR1RVVcXSpUvjkksuKd7kAEBZ+kL3fDQ1NUVERO/evSMior6+PhobG2PcuHFt++Tz+Rg9enSsW7duj+/R0tISzc3N7RYA4MC13/GRZVlMnz49Ro4cGYMHD46IiMbGxoiIqKqqardvVVVV22ufNmfOnKisrGxb+vXrt78jAQBlYL/j4/LLL49nn302/vjHP+72Wi6Xa7eeZdlu2z4xc+bMaGpqalsaGhr2dyQAoAwUdM/HJ6644oq47777Yu3atXHYYYe1ba+uro6I/14Bqampadu+bdu23a6GfCKfz0c+n9+fMQCAMlTQlY8sy+Lyyy+PZcuWxcMPPxwDBw5s9/rAgQOjuro6Vq1a1bbtgw8+iDVr1sSIESOKMzEAUNYKuvIxderUWLp0afzpT3+KioqKtvs4Kisro0ePHpHL5WLatGkxe/bsGDRoUAwaNChmz54dPXv2jEmTJnXI/wAAUF4Kio9FixZFRMSYMWPabV+yZElMmTIlIiKuvfba2LVrV1x22WXxzjvvxPDhw+PBBx+MioqKogwMAJS3guIjy7K97pPL5aKuri7q6ur2dyYA4ADm2S4AQFLiAwBISnwAAEmJDwAgKfEBACQlPgCApMQHAJCU+AAAkhIfAEBS4gMASEp8AABJiQ8AICnxAQAkJT4AgKTEBwCQlPgAAJISHwBAUuIDAEhKfAAASYkPACAp8QEAJCU+AICkxAcAkJT4AACSEh8AQFLiAwBISnwAAEmJDwAgKfEBACQlPgCApMQHAJCU+AAAkhIfAEBS4gMASEp8AABJiQ8AICnxAQAkJT4AgKTEBwCQlPgAAJISHwBAUuIDAEhKfAAASYkPACAp8QEAJCU+AICkxAcAkJT4AACSEh8AQFLiAwBISnwAAEmJDwAgqa6lHgCgHBw+Y0WpRyjY1rmnlnoE2CNXPgCApMQHAJCU+AAAkhIfAEBS4gMASEp8AABJiQ8AICnxAQAkJT4AgKTEBwCQVMHxsXbt2pgwYULU1tZGLpeL5cuXt3t9ypQpkcvl2i0nnnhiseYFAMpcwfGxc+fOGDp0aCxcuPAz9znllFPizTffbFv+8pe/fKEhAYADR8EPlhs/fnyMHz/+c/fJ5/NRXV29T+/X0tISLS0tbevNzc2FjgQAlJEOuedj9erV0bdv3zjyyCPjxz/+cWzbtu0z950zZ05UVla2Lf369euIkQCATqLo8TF+/Pj4wx/+EA8//HDcfPPNsX79+vjud7/b7urG/5o5c2Y0NTW1LQ0NDcUeCQDoRAr+2GVvzj333Lb/Hjx4cAwbNiwGDBgQK1asiIkTJ+62fz6fj3w+X+wxAIBOqsN/1bampiYGDBgQW7Zs6ehDAQBloMPj4+23346Ghoaoqanp6EMBAGWg4I9dduzYES+//HLben19fWzatCl69+4dvXv3jrq6ujjrrLOipqYmtm7dGj/72c+iT58+ceaZZxZ1cACgPBUcH0899VSMHTu2bX369OkRETF58uRYtGhRPPfcc3HnnXfGu+++GzU1NTF27Ni45557oqKionhTAwBlq+D4GDNmTGRZ9pmvr1y58gsNBAAc2DzbBQBISnwAAEmJDwAgKfEBACQlPgCApMQHAJCU+AAAkhIfAEBS4gMASEp8AABJiQ8AICnxAQAkJT4AgKTEBwCQlPgAAJISHwBAUuIDAEhKfAAASYkPACAp8QEAJCU+AICkxAcAkJT4AACSEh8AQFLiAwBISnwAAEmJDwAgKfEBACQlPgCApMQHAJCU+AAAkhIfAEBS4gMASEp8AABJiQ8AICnxAQAkJT4AgKTEBwCQlPgAAJISHwBAUuIDAEhKfAAASYkPACAp8QEAJCU+AICkxAcAkJT4AACSEh8AQFLiAwBISnwAAEmJDwAgKfEBACQlPgCApMQHAJCU+AAAkhIfAEBS4gMASEp8AABJiQ8AICnxAQAkJT4AgKTEBwCQVMHxsXbt2pgwYULU1tZGLpeL5cuXt3s9y7Koq6uL2tra6NGjR4wZMyY2b95crHkBgDJXcHzs3Lkzhg4dGgsXLtzj6/PmzYv58+fHwoULY/369VFdXR0nn3xybN++/QsPCwCUv66FfsH48eNj/Pjxe3wty7JYsGBBzJo1KyZOnBgREXfccUdUVVXF0qVL45JLLvli0wIAZa+o93zU19dHY2NjjBs3rm1bPp+P0aNHx7p16/b4NS0tLdHc3NxuAQAOXEWNj8bGxoiIqKqqare9qqqq7bVPmzNnTlRWVrYt/fr1K+ZIAEAn0yG/7ZLL5dqtZ1m227ZPzJw5M5qamtqWhoaGjhgJAOgkCr7n4/NUV1dHxH+vgNTU1LRt37Zt225XQz6Rz+cjn88XcwwAoBMr6pWPgQMHRnV1daxatapt2wcffBBr1qyJESNGFPNQAECZKvjKx44dO+Lll19uW6+vr49NmzZF7969o3///jFt2rSYPXt2DBo0KAYNGhSzZ8+Onj17xqRJk4o6OABQngqOj6eeeirGjh3btj59+vSIiJg8eXL8/ve/j2uvvTZ27doVl112WbzzzjsxfPjwePDBB6OioqJ4UwMAZavg+BgzZkxkWfaZr+dyuairq4u6urovMhcAcIDybBcAICnxAQAkJT4AgKTEBwCQlPgAAJISHwBAUuIDAEhKfAAASYkPACAp8QEAJCU+AICkxAcAkJT4AACSEh8AQFLiAwBISnwAAEmJDwAgKfEBACQlPgCApMQHAJCU+AAAkhIfAEBSXUs9AAemw2esKPUIBds699RSjwDwpeDKBwCQlPgAAJISHwBAUuIDAEhKfAAASYkPACAp8QEAJCU+AICkxAcAkJT4AACSEh8AQFLiAwBISnwAAEmJDwAgqa6lHgAAPnH4jBWlHqFgW+eeWuoRyo4rHwBAUuIDAEhKfAAASYkPACAp8QEAJCU+AICkxAcAkJT4AACSEh8AQFLiAwBISnwAAEmJDwAgKfEBACQlPgCApMQHAJCU+AAAkhIfAEBS4gMASEp8AABJiQ8AICnxAQAkJT4AgKTEBwCQlPgAAJIqenzU1dVFLpdrt1RXVxf7MABAmeraEW96zDHHxEMPPdS23qVLl444DABQhjokPrp27epqBwCwRx1yz8eWLVuitrY2Bg4cGOedd1688sorn7lvS0tLNDc3t1sAgANX0eNj+PDhceedd8bKlSvj9ttvj8bGxhgxYkS8/fbbe9x/zpw5UVlZ2bb069ev2CMBAJ1I0eNj/PjxcdZZZ8WQIUPi//7v/2LFihUREXHHHXfscf+ZM2dGU1NT29LQ0FDskQCATqRD7vn4XwcffHAMGTIktmzZssfX8/l85PP5jh4DAOgkOvzf+WhpaYkXXnghampqOvpQAEAZKHp8XH311bFmzZqor6+PJ554Is4+++xobm6OyZMnF/tQAEAZKvrHLq+//nqcf/758dZbb8XXvva1OPHEE+Pxxx+PAQMGFPtQAEAZKnp83H333cV+SwDgAOLZLgBAUuIDAEhKfAAASYkPACAp8QEAJCU+AICkxAcAkJT4AACS6vAHy0G5OHzGilKPULCtc08t9QgABXPlAwBISnwAAEmJDwAgKfEBACQlPgCApMQHAJCU+AAAkhIfAEBS4gMASEp8AABJiQ8AICnxAQAkJT4AgKQ81bYMlOPTVgHgs7jyAQAkJT4AgKTEBwCQlPgAAJISHwBAUuIDAEhKfAAASYkPACAp8QEAJCU+AICkxAcAkJT4AACSEh8AQFLiAwBIqmupBwCAcnb4jBWlHqFgW+eeWtLju/IBACQlPgCApMQHAJCU+AAAkhIfAEBS4gMASEp8AABJiQ8AICnxAQAkJT4AgKTEBwCQlPgAAJISHwBAUuIDAEiqa6kHAL5cyvHx4+XKuaazcuUDAEhKfAAASYkPACAp8QEAJCU+AICkxAcAkJT4AACSEh8AQFLiAwBISnwAAEl1WHzceuutMXDgwOjevXscf/zx8eijj3bUoQCAMtIh8XHPPffEtGnTYtasWfH000/Hd77znRg/fny89tprHXE4AKCMdMiD5ebPnx8//OEP40c/+lFERCxYsCBWrlwZixYtijlz5rTbt6WlJVpaWtrWm5qaIiKiubm5I0aL1pb3OuR9oRQ66s9JR/JnEEqvI/7u+OQ9syzb+85ZkbW0tGRdunTJli1b1m77lVdemY0aNWq3/W+44YYsIiwWi8VisRwAS0NDw15boehXPt566634+OOPo6qqqt32qqqqaGxs3G3/mTNnxvTp09vWW1tb4z//+U8ceuihkcvlij1ep9fc3Bz9+vWLhoaG6NWrV6nHKVvOY3E4j8XhPBaH81gcHXUesyyL7du3R21t7V737ZCPXSJit3DIsmyPMZHP5yOfz7fbdsghh3TUWGWjV69e/nAVgfNYHM5jcTiPxeE8FkdHnMfKysp92q/oN5z26dMnunTpsttVjm3btu12NQQA+PIpenx069Ytjj/++Fi1alW77atWrYoRI0YU+3AAQJnpkI9dpk+fHhdeeGEMGzYsvv3tb8fixYvjtddei0svvbQjDndAyefzccMNN+z2URSFcR6Lw3ksDuexOJzH4ugM5zGXZfvyOzGFu/XWW2PevHnx5ptvxuDBg+OWW26JUaNGdcShAIAy0mHxAQCwJ57tAgAkJT4AgKTEBwCQlPgAAJISH53Iv//977jgggvi0EMPjZ49e8axxx4bGzZsKPVYZeWjjz6Kn//85zFw4MDo0aNHHHHEEfHLX/4yWltbSz1ap7Z27dqYMGFC1NbWRi6Xi+XLl7d7PcuyqKuri9ra2ujRo0eMGTMmNm/eXJphO7HPO48ffvhhXHfddTFkyJA4+OCDo7a2Ni666KJ44403SjdwJ7W378f/dckll0Qul4sFCxYkm69c7Mt5fOGFF+K0006LysrKqKioiBNPPDHJE+jFRyfxzjvvxEknnRQHHXRQ/PWvf43nn38+br75Zv/UfIFuuummuO2222LhwoXxwgsvxLx58+LXv/51/Pa3vy31aJ3azp07Y+jQobFw4cI9vj5v3ryYP39+LFy4MNavXx/V1dVx8sknx/bt2xNP2rl93nl87733YuPGjXH99dfHxo0bY9myZfHSSy/FaaedVoJJO7e9fT9+Yvny5fHEE0/s07NEvoz2dh7/9a9/xciRI+Poo4+O1atXxzPPPBPXX399dO/eveOH++LPsaUYrrvuumzkyJGlHqPsnXrqqdnFF1/cbtvEiROzCy64oEQTlZ+IyO6999629dbW1qy6ujqbO3du27b3338/q6yszG677bYSTFgePn0e9+TJJ5/MIiJ79dVX0wxVhj7rPL7++uvZ17/+9ewf//hHNmDAgOyWW25JPls52dN5PPfcc0v2d6MrH53EfffdF8OGDYsf/OAH0bdv3zjuuOPi9ttvL/VYZWfkyJHxt7/9LV566aWIiHjmmWfisccei+9///slnqx81dfXR2NjY4wbN65tWz6fj9GjR8e6detKOFn5a2pqilwu5wpngVpbW+PCCy+Ma665Jo455phSj1OWWltbY8WKFXHkkUfG9773vejbt28MHz78cz/iKibx0Um88sorsWjRohg0aFCsXLkyLr300rjyyivjzjvvLPVoZeW6666L888/P44++ug46KCD4rjjjotp06bF+eefX+rRytYnD4n89IMhq6qqdnuAJPvu/fffjxkzZsSkSZM8obVAN910U3Tt2jWuvPLKUo9StrZt2xY7duyIuXPnximnnBIPPvhgnHnmmTFx4sRYs2ZNhx+/Q57tQuFaW1tj2LBhMXv27IiIOO6442Lz5s2xaNGiuOiii0o8Xfm455574q677oqlS5fGMcccE5s2bYpp06ZFbW1tTJ48udTjlbVcLtduPcuy3baxbz788MM477zzorW1NW699dZSj1NWNmzYEL/5zW9i48aNvv++gE9uwj/99NPjqquuioiIY489NtatWxe33XZbjB49ukOP78pHJ1FTUxPf+ta32m375je/meSu4wPJNddcEzNmzIjzzjsvhgwZEhdeeGFcddVVMWfOnFKPVraqq6sjIna7yrFt27bdroawdx9++GGcc845UV9fH6tWrXLVo0CPPvpobNu2Lfr37x9du3aNrl27xquvvho//elP4/DDDy/1eGWjT58+0bVr15L93BEfncRJJ50UL774YrttL730UgwYMKBEE5Wn9957L77ylfbf1l26dPGrtl/AwIEDo7q6OlatWtW27YMPPog1a9bEiBEjSjhZ+fkkPLZs2RIPPfRQHHrooaUeqexceOGF8eyzz8amTZvaltra2rjmmmti5cqVpR6vbHTr1i1OOOGEkv3c8bFLJ3HVVVfFiBEjYvbs2XHOOefEk08+GYsXL47FixeXerSyMmHChLjxxhujf//+ccwxx8TTTz8d8+fPj4svvrjUo3VqO3bsiJdffrltvb6+PjZt2hS9e/eO/v37x7Rp02L27NkxaNCgGDRoUMyePTt69uwZkyZNKuHUnc/nncfa2to4++yzY+PGjfHnP/85Pv7447arSb17945u3bqVauxOZ2/fj5+OtoMOOiiqq6vjqKOOSj1qp7a383jNNdfEueeeG6NGjYqxY8fGAw88EPfff3+sXr2644crye/YsEf3339/Nnjw4Cyfz2dHH310tnjx4lKPVHaam5uzn/zkJ1n//v2z7t27Z0cccUQ2a9asrKWlpdSjdWqPPPJIFhG7LZMnT86y7L+/bnvDDTdk1dXVWT6fz0aNGpU999xzpR26E/q881hfX7/H1yIie+SRR0o9eqeyt+/HT/Ortnu2L+fxd7/7XfaNb3wj6969ezZ06NBs+fLlSWbLZVmWdXziAAD8l3s+AICkxAcAkJT4AACSEh8AQFLiAwBISnwAAEmJDwAgKfEBACQlPgCApMQHAJCU+AAAkvp/gCOHB3SUsk4AAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "\n",
    "plt.hist(list(hits.values()),range=(5,16))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "222\n"
     ]
    }
   ],
   "source": [
    "# Get index of model given gene name\n",
    "gene = 'GR'\n",
    "\n",
    "for i, motif in enumerate(MOTIFCENTRAL):\n",
    "    if gene in motif['metadata']['factors'][0]['gene_symbol']:\n",
    "        print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'modelSettings': {'letterOrder': 'ACGT',\n",
       "  'bindingModes': [{'fitLogActivity': True,\n",
       "    'flankLength': 0,\n",
       "    'singleStrand': False,\n",
       "    'size': 18,\n",
       "    'dinucleotideDistance': 0,\n",
       "    'componentName': 'Binding mode 1',\n",
       "    'positionBias': False,\n",
       "    'modifications': []}],\n",
       "  'letterComplement': 'C-G,A-T'},\n",
       " 'coefficients': {'bindingModes': [{'mononucleotide': [-0.8892676406380744,\n",
       "     -0.17695740932441772,\n",
       "     0,\n",
       "     -0.1906680224545454,\n",
       "     0,\n",
       "     -0.2736090466940002,\n",
       "     -0.32218982223937154,\n",
       "     -0.5090409965771627,\n",
       "     -1.094321502733333,\n",
       "     -1.2095045360613295,\n",
       "     0,\n",
       "     -1.4175601915336813,\n",
       "     -0.5818222024313133,\n",
       "     0,\n",
       "     -1.5835424380311665,\n",
       "     -0.9661045270704722,\n",
       "     -2.2910094805489507,\n",
       "     -2.1748712561584336,\n",
       "     0,\n",
       "     -1.8172344308689308,\n",
       "     -1.9136461442114072,\n",
       "     0,\n",
       "     -2.0211043009272736,\n",
       "     -2.53754491416455,\n",
       "     -2.8599097340573456,\n",
       "     0,\n",
       "     -3.184751722068996,\n",
       "     -2.793320371083581,\n",
       "     -0.4245389512208142,\n",
       "     0,\n",
       "     -1.7128805796494966,\n",
       "     -1.9087395102605136,\n",
       "     -2.115705973271286,\n",
       "     0,\n",
       "     -3.1693761079291676,\n",
       "     -0.7547820853493548,\n",
       "     -3.010953759565745,\n",
       "     0,\n",
       "     -3.20828955528704,\n",
       "     -2.8090419627375667,\n",
       "     -2.0195026519080574,\n",
       "     -1.88314575261397,\n",
       "     -2.246483065403515,\n",
       "     0,\n",
       "     0,\n",
       "     -1.5585446110615226,\n",
       "     -0.6210689829925347,\n",
       "     -1.010901822737404,\n",
       "     -1.9002916392889784,\n",
       "     -0.8337049296699224,\n",
       "     0,\n",
       "     -1.1797822549447154,\n",
       "     -0.738164653253488,\n",
       "     0,\n",
       "     -1.2307227179571,\n",
       "     -0.012353521978297088,\n",
       "     -3.0011332525258387,\n",
       "     -1.9569920594761632,\n",
       "     0,\n",
       "     -1.7342850660196305,\n",
       "     -1.3594885117441446,\n",
       "     -1.9099101071587148,\n",
       "     0,\n",
       "     -1.114237929416988,\n",
       "     -0.6645156362564677,\n",
       "     -0.2762126973083161,\n",
       "     -1.0553283505825277,\n",
       "     0,\n",
       "     -0.4349019280563624,\n",
       "     0,\n",
       "     -0.3053046379897246,\n",
       "     -0.6884150155906488],\n",
       "    'activity': [[0]],\n",
       "    'dinucleotide': [],\n",
       "    'positionBias': [],\n",
       "    'modifications': []}]},\n",
       " 'metadata': {'fit_id': 12715,\n",
       "  'factors': [{'gene_symbol': 'CTCF',\n",
       "    'gene_name': 'CCCTC-binding factor',\n",
       "    'tax_id': 9606,\n",
       "    'annotation': [['ENSEMBL_GENE', 'ENSG00000102974'],\n",
       "     ['ENTREZ', '10664'],\n",
       "     ['HGNC', '13723'],\n",
       "     ['PFAM', 'PF00096'],\n",
       "     ['PFAM', 'PF13909'],\n",
       "     ['SWISS_PROT', 'P49711']]}],\n",
       "  'experiments': [['Isakova2017', 'CTCF_r1'],\n",
       "   ['Isakova2017', 'CTCF_r2'],\n",
       "   ['Isakova2017', 'CTCF_r3'],\n",
       "   ['Jolma2013', 'CTCF_ESAJ_TAGCGA20NGCT']]}}"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MOTIFCENTRAL[222]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[-0.038763142589396615,\n",
       " -0.13191122334958205,\n",
       " 0,\n",
       " -0.1628007128506166,\n",
       " -0.1713360048105077,\n",
       " 0,\n",
       " -0.06891111980367448,\n",
       " -0.14441890323581408,\n",
       " -0.6913516948517531,\n",
       " -0.4562045373322672,\n",
       " 0,\n",
       " -0.000557632265413055,\n",
       " 0,\n",
       " -1.751194866226065,\n",
       " -0.430016165775819,\n",
       " -2.1640772471029788,\n",
       " -3.3905916431300236,\n",
       " -3.402173256954452,\n",
       " -3.5193330959607305,\n",
       " 0,\n",
       " -3.233472293182503,\n",
       " -4.155789300356583,\n",
       " -1.675808295105874,\n",
       " 0,\n",
       " -0.7294225472925886,\n",
       " -3.451088089019582,\n",
       " 0,\n",
       " -1.7419515768521707,\n",
       " -2.4330022573130483,\n",
       " 0,\n",
       " -2.4716176379839556,\n",
       " -1.626498721615297,\n",
       " -1.370751947914326,\n",
       " -2.6242176632930985,\n",
       " 0,\n",
       " -2.48343206040465,\n",
       " -1.3529142815463429,\n",
       " 0,\n",
       " -3.0060066045186318,\n",
       " -0.5921247288245386,\n",
       " 0,\n",
       " -1.0570731289883604,\n",
       " -3.4181294651806073,\n",
       " -3.5335692319182623,\n",
       " -0.04181456473679912,\n",
       " 0,\n",
       " -2.560380716174281,\n",
       " -2.9629212117346038,\n",
       " -0.3675887706865124,\n",
       " -0.3898058260833803,\n",
       " -0.11692407308984709,\n",
       " 0,\n",
       " 0,\n",
       " -0.4345145186557877,\n",
       " -0.9350357423634765,\n",
       " -0.469741598589116]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MOTIFCENTRAL[13][\"coefficients\"][\"bindingModes\"][0]['mononucleotide']"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SelexXGenome",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
